[mlir][debug] Convert DbgIntrOp to DebugRecords directly.

abidh opened the pull request:
Currently the `DbgDeclareOP/DbgValueOP/DbgLabelOp` are first converted to llvm debug intrinsics which are later translated to debug records by a call of `convertToNewDbgValues`. This is not only inefficient but also makes the code that works on intermediate IR unnecessarily complicated. The debug intrinsics are also being phased out. This PR converts these Ops directly to debug records.

The conversion is relatively simple but there is a bit of code repetition due to how the APIs in the `DIBuilders` are named. There are few `cast<>` which I would like to do without but could not see a good way around them. Any suggestions welcome here. Also noticed that `DISubprogramAttr` is inherited from `DIScopeAttr` while in llvm, the `DISubprogram` inherits from `DILocalScope`. I am going to fix this separately and then we could use `FusedLocWith<LLVM::DILocalScopeAttr>` and cast to `DILocalScope` will be much safer.

As the output remains the same, the existing tests cover this change. I also ran the `GDB` tests with flang and there was no regression.

abidh committed:
[mlir] Convert to DebugRecords directly.

Currently the DbgIntrOp are first converted to llvm intrinsics
which are laterr translated to debug records by a call of
convertToNewDbgValues. This is not only inefficient but also makes
the code that works on intermediate IR unnecessarily complicated. This
PR converts the DbgIntrOp directly to debug records.

The conversion is relatively simple but there is a bit of code
repetition due to how the APIs in the DIBuilders are named. Also noticed
that DISubprogramAttr is inherited from DIScopeAttr while in llvm, the
DISubprogram inherits from DILocalScope. I am going to fix this
separately and then we could use FusedLocWith<LLVM::DILocalScopeAttr>
and cast to <DILocalScope> will be much safer.

SLTozer commented on mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td:
```diff
@@ -711,6 +695,28 @@ def LLVM_DbgDeclareOp : LLVM_DbgIntrOp<"dbg.declare", "addr", [
     DefaultValuedAttr<LLVM_DIExpressionAttr, "{}">:$locationExpr
   );
+  let llvmBuilder = [{
+    // Debug intrinsics without debug locations are invalid.
```
May as well update the terminology while passing by.
```suggestion
    // Debug records without debug locations are invalid.
```

SLTozer commented on mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td:
```diff
@@ -721,6 +727,29 @@ def LLVM_DbgValueOp : LLVM_DbgIntrOp<"dbg.value", "value",
     DefaultValuedAttr<LLVM_DIExpressionAttr, "{}">:$locationExpr
   );
+  let llvmBuilder = [{
+    // Debug intrinsics without debug locations are invalid.
```
Ditto above.
```suggestion
    // Debug records without debug locations are invalid.
```

SLTozer reviewed (commented):
Can't speak to the MLIR logic, but the DIBuilder usage LGTM (also looks like we missed a spot, leaving the debug value creation function as `insertDbgValueIntrinsic`!) - thanks for updating this!

gysit commented on mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td:
```diff
@@ -721,6 +727,29 @@ def LLVM_DbgValueOp : LLVM_DbgIntrOp<"dbg.value", "value",
+    if(!builder.getCurrentDebugLocation())
+      return success();
+    llvm::Function *parentFn = builder.GetInsertBlock()->getParent();
+    llvm::DILocalScope *scope;
+    if (auto scopeLoc =
+            opInst.getLoc()->findInstanceOf<FusedLocWith<LLVM::DIScopeAttr>>())
+      scope = llvm::cast<llvm::DILocalScope>(
+          moduleTranslation.translateDebugInfo(scopeLoc.getMetadata()));
+    else
+      scope = parentFn->getSubprogram();
```
Since it is not possible anymore to put all the builder code in the debug intrinsic base class there is quite a bit of code duplication. Can you factor out some of this code into a helper function(s) placed in LLVMDialect.cpp? One thing that seems doable is to extract a helper for the scope computation but there may be other options as well. I would place it somewhere to the other intrinsic functionality that has been factored out (e.g. here https://github.com/llvm/llvm-project/blob/3054e06c33ac8977b355bf2cff14617a342dbce9/mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp#L4168).

That should reduce code duplication and the amount of C++ embedded in tablegen.

abidh replied:
Thanks for your suggestion. I have moved the code in a separate function. I initially used the location that you suggested for the new function but it would have required to add its declaration in `LLVMDialect.h` and then at least a forward declaration of `ModuleTranslation` there. It seemed like too much churn for a local helper. I have instead added in `LLVMToLLVMIRTranslation.cpp`.

I also changed the code to use dyn_cast.

gysit commented on mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td:
```diff
@@ -721,6 +727,29 @@ def LLVM_DbgValueOp : LLVM_DbgIntrOp<"dbg.value", "value",
+      scope = parentFn->getSubprogram();
+
+    llvm::Module *module = builder.GetInsertBlock()->getModule();
+    llvm::DIBuilder DB(*module);
```
```suggestion
    llvm::DIBuilder debugInfoBuilder(*module);
```
nit: local variables should start with a lower case letter. Let's maybe use debugInfoBuilder or similar?

gysit reviewed (commented):
The change makes sense!

Let's try to factor out some helpers to reduce the code duplication.

The casts are probably unavoidable. If there a place where the cast could fail we may want to use dyn_cast and return failure to have a better error handling. However, it sounds like refactoring the inheritance to match LLVM will fix the critical case?

andykaylor reviewed (commented):
Can you add a test for the new behavior?

abidh committed:
Fix review comments.

1. Refactor some code into a separate function to avoid duplication.
2. Some name and comments update.

abidh commented:
> Can you add a test for the new behavior?

As I wrote in the PR description, the output remains the same. It is just that intermediate use of llvm intrinsics have been removed. So all the existing tests check this functionality.

gysit commented on mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td:
```diff
@@ -673,22 +673,6 @@ def LLVM_CoroPromiseOp : LLVM_IntrOp<"coro.promise", [], [], [], 1> {
 
 class LLVM_DbgIntrOp<string name, string argName, list<Trait> traits = []>
```
Yeah I would stick to `LLVM_DbgIntrOp` since they are effectively still intrinsics in MLIR.

andykaylor replied:
In what sense are they intrinsics in MLIR? I didn't think MLIR had a concept of intrinsics beyond having a way to represent calls to LLVM IR intrinsics.

This is obviously a minor point and I don't have a problem with the change going in either way. I just want to make sure I'm not missing something.

gysit replied:
You are correct that MLIR does not have the concept of an intrinsic. However, LLVM dialect still models intrinsics, using operations, and it doesn't have the equivalent of debug records. Ideally, a debug record would not be an MLIR operation, but some sort of Location attribute attached to an operation. Unfortunately, MLIR Locations cannot depend on values which would be necessary for debug records if I understand them correctly. 

TD;LR we will need to come up with a modeling for debug records. Ideally, this would be done with MLIR's Location infrastructure, but I fear we will have to stick to operations.

andykaylor replied:
Yes, there's definitely a difficulty mapping debug records to MLIR constructs. I'm fine with this implementation if it's just seen as a temporary step along the way to a better handling of debug records, though this artificial link to intrinsic calls makes me uneasy.

gysit replied:
Yes I agree that this has to change. We just miss volunteers to implement this. There is a ticket for this (https://github.com/llvm/llvm-project/issues/129121)

gysit reviewed (approved):
Thanks for addressing the comments.

LGTM from my end modulo last nit.

abidh committed:
Update mlir/lib/Target/LLVMIR/Dialect/LLVMIR/LLVMToLLVMIRTranslation.cpp

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

abidh commented on mlir/include/mlir/Dialect/LLVMIR/LLVMIntrinsicOps.td:
```diff
@@ -711,6 +695,28 @@ def LLVM_DbgDeclareOp : LLVM_DbgIntrOp<"dbg.declare", "addr", [
+  let llvmBuilder = [{
+    // Debug intrinsics without debug locations are invalid.
+    if(!builder.getCurrentDebugLocation())
+      return success();
```
I was looking at changing this to an error and noticed that the behavior is intentional. It was requested in https://github.com/llvm/llvm-project/issues/60222 and implemented in https://github.com/llvm/llvm-project/pull/70643.
